### PR TITLE
Added JWT tokens capability

### PR DESCRIPTION
**Pull request by UXabre**
This PR adds the capability to use (a)symmetric JWTs as token for retrieving connection details. The only argument needed, is the public key or shared secret file for verification. 

This way, the management of tokens to hosts becomes very manageable and quite secure.

I've also added some tests to confirm its functionality. Needless to say that the public/private keys displayed here are only for testing and mustn't be used in production

Let me know what you guys think!

### CODE REVIEW

**Comment by CendioOssman** [comment]
Looks good, but I'd like to avoid a hard requirement on PyJWT. So could you omit the change to `setup.py`? You can put it in `test-requirements.txt` instead for the tests to succeed.

As for the tests failing, there seems to be some issues with the `cryptography` module. Is that really needed? I don't see any imports for it.

**Comment by UXabre** [comment]
Hey @CendioOssman, thanks for reviewing! Unfortunately, the pyjwt is needed for this code to work (all testing aside), the import happens a bit further down the line: https://github.com/novnc/websockify/pull/372/files#diff-f7cac2852d62df1cca378dcc6467212bR97

How would we best proceed with that?

On your second note, regarding cryptography lib, it's needed to support all security hashes in pyjwt. So, they're kinda "married" if we want to support the entire spectrum of jwt hashes.

File: websockify/token_plugins.py [hunk 1]
```diff
@@ -87,3 +87,19 @@ class JSONTokenApi(BaseTokenAPI):
...
+
+class JWTTokenApi(BasePlugin):
+    # source is a JWT-token, with hostname and port included
+    # this JWT is validated against a assymetric key, if provided else, nothing is checked!
+
+    def lookup(self, token):
+        import jwt
+
+        try:
+            secret = open(self.source).read()
```
Comment (ekohl) [inline websockify/token_plugins.py]: Using `with` automatically closes the file ensures you don't leak file handles:
```python
with open(self.source) as secret_file:
    secret = secret_file.read()
```

File: websockify/token_plugins.py [hunk 2]
```diff
@@ -87,3 +87,19 @@ class JSONTokenApi(BaseTokenAPI):
...
+            parsed = jwt.decode(token, secret, algorithms=['RS256', 'RS384', 'RS512', 'HS256', 'HS384', 'HS512', 'ES256', 'ES384', 'ES512', 'PS256', 'PS384', 'PS512'])
+
+            return (parsed['host'], parsed['port'])
+        except:
+            return None
```
Comment (ekohl) [inline websockify/token_plugins.py]: A newline at the end is generally preferred.

File: websockify/token_plugins.py [hunk 3]
```diff
@@ -87,3 +87,19 @@ class JSONTokenApi(BaseTokenAPI):
...
+
+    def lookup(self, token):
+        import jwt
+
+        try:
+            secret = open(self.source).read()
+            parsed = jwt.decode(token, secret, algorithms=['RS256', 'RS384', 'RS512', 'HS256', 'HS384', 'HS512', 'ES256', 'ES384', 'ES512', 'PS256', 'PS384', 'PS512'])
+
+            return (parsed['host'], parsed['port'])
+        except:
```
Comment (ekohl) [inline websockify/token_plugins.py]: It would be nice if there was a way to log the exception that happened so you can debug things.

**Comment by UXabre** [comment]
> Does this add the ability to pass just a token to the websocket connection to specify the host/port/credentials? If this is the case, would be possible to generate the token on server side, pass it to the novnc client which passes it to websocksify but in a way that the token is non readable from the client (by readable I mean the user can't just go to jwt.io and read the token host/port/credentials)?

Hey @alex88 , yet, it is possible to provide a whole heap of information to the receiving end, including port and address. The credentials are not passed (as websockify basically translates from a "regular" TCP socket to a websocket, it doesn't know about the session/application/presentation layer, credentials are thus passed somewhere higher up the chain). However, in this case, I've only added support for JWS', what you seem to describe is JWE. I'd have to investigate but I'd at least would have to switch JWT library (and read up if it has no vulnerabilities on that part).

Given this, one could probably debate if then a port and address are worth keeping a secret to begin with? (credentials would be a different cake)

**Comment by CendioOssman** [comment]
People have already complained about numpy being a hard requirement, so I'm not keen on adding more. Especially requirements that will not be used by most users. So I'm afraid this will have to be an optional thing.

**Comment by ekohl** [comment]
> Given this, one could probably debate if then a port and address are worth keeping a secret to begin with? (credentials would be a different cake)

Hostnames/addresses are internal details that you might not want to leak.

**Comment by CendioOssman** [comment]
> @CendioOssman, okay, what would the best way of making this optional, or do I just add it to the test-requirements.txt document? :-)

I don't think we have anything better than that, no. Possibly you could catch import errors and print something more useful.

**Comment by CendioOssman** [comment]
The tests are failing, but otherwise this looks good to me. So get those working and then we can merge this.

File: websockify/token_plugins.py [hunk 4]
```diff
@@ -87,3 +87,45 @@ class JSONTokenApi(BaseTokenAPI):
...
         resp_json = resp.json()
         return (resp_json['host'], resp_json['port'])
+
+
+class JWTTokenApi(BasePlugin):
+    # source is a JWT-token, with hostname and port included
+    # Both JWS as JWE tokens are accepted. With regards to JWE tokens, the key is re-used for both validation and decryption.
+
+    def lookup(self, token):
+        from jwcrypto import jwt
```
Comment (ekohl) [inline websockify/token_plugins.py]: I'd catch `ImportError` here as well and handle it with a proper error message

Comment (UXabre) [reply websockify/token_plugins.py]: Oh, yes, should've done this right away, will fix! Thanks!

**Comment by UXabre** [comment]
> The tests are failing, but otherwise this looks good to me. So get those working and then we can merge this.

I noticed the tests are failing (at least for python > 2.7) because setuptools seems to be outdated? To mee, this looks more like a build infrastructure issue; no? Is this something that we could update? Or, is there a way I could enforce the upgrade via this commit?

Seems 2.6 is a different can of worms, will need to investigate what's going on there... perhaps I skip support for this feature on python <2.7?

File: websockify/token_plugins.py (hunk 4 above)
Comment (UXabre) [reply websockify/token_plugins.py]: Done and done!

File: websockify/token_plugins.py [hunk 5]
```diff
@@ -87,3 +87,49 @@ class JSONTokenApi(BaseTokenAPI):
...
+
+    def lookup(self, token):
+        try:
+            from jwcrypto import jwt
+            import json
+
+            key = jwt.JWK()
+            
+            try:
+                with open(self.source) as key_file:
```
Comment (cbosdo) [inline websockify/token_plugins.py]: I would open the file in byte mode here. Nothing guarantees the key will only contain valid utf-8 characters. Unless you explicitely require the key to be hex-encoded, which is not the case here.

Comment (UXabre) [reply websockify/token_plugins.py]: Hmm, never thought of that. Will add 'rb' as mode :-) Thank you for your input!

**Comment by CendioOssman** [comment]
> > 
> 
> I noticed the tests are failing (at least for python > 2.7) because setuptools seems to be outdated? To mee, this looks more like a build infrastructure issue; no? Is this something that we could update? Or, is there a way I could enforce the upgrade via this commit?
> 

Maybe. The setup is in the `.travis.yml` file. You could try playing around with updating things there.

> Seems 2.6 is a different can of worms, will need to investigate what's going on there... perhaps I skip support for this feature on python <2.7?

That is fine by me.

**Comment by UXabre** [comment]
Allright, another day, another update!
I've managed to make the build work, but, as expected, I had to make some exceptions for the "ancient" python versions.

(not sure if we should still support Python2 at all actually because it'll soon be "retired": https://pythonclock.org/ ; but that's totally outside of the scope of this PR :p)

**Comment by CendioOssman** [comment]
Awesome. Thanks for all your work here.

**Pull request merged**